### ISSUE TITLE
Generalize auxiliary trace building logic

### ISSUE THREAD

**Issue by Al-Kindi-0** [facebook/winterfell#257]
With #240 almost done via #247 , it makes sense to discuss how we can make use of these recent updates in the specific context of LogUp GKR.

# Current
To recap, at a high level, the current workflow between the (STARK) prover and verifier can be described using the following simplified diagram
![original_aux_trace](https://github.com/facebook/winterfell/assets/82364884/0d601dce-e4ff-4ca9-a253-c8c34d230796)

In the above
1.  $\mathsf{f_i}$ is a polynomial encoding the $i$-th main trace column and $\mathsf{commit(f_i)}$ is its commitment.
2. $\alpha := \left(\alpha_i\right)_i$ is a vector of random values used in reducing global constraints to local ones e.g., multi-set checks.
3.  $\mathsf{f_i^{'}}$ is a polynomial encoding the $i$-th auxiliary trace column and $\mathsf{commit(f_i^{'})}$ is its commitment.
4. $\beta := \left(\beta_i\right)_i$ is a vector of random values used in combining constraints in order to construct the constraint composition polynomial.
5.  $\mathsf{g_i}$ is a polynomial encoding the $i$-th segment polynomial of the constraint composition polynomial and $\mathsf{commit(g_i)}$ is its commitment.
6. $\mathsf{z}$ is the OOD evaluation point.
7. $\mathsf{v_i}$ are the evaluations of all the previous oracles (i.e., committed polynomials) at $\mathsf{z}$ and potentially their shifts.
8. $\gamma := \left(\gamma_i\right)_i$ is a vector of random values used in constructing the DEEP polynomial i.e., the input to FRI.
9. $\mathsf{h}$ is a polynomial encoding the DEEP polynomial and $\mathsf{commit(h)}$ is its commitment.

Note that FRI is invoked at the end in order to show low degree-ness as well as prove the openings. FRI is itself an IOP and thus is composed of multiple rounds of the prover sending commitments to polynomials and the verifier replying with randomness. The verifier concludes by doing a few checks on the messages it received from the prover and the randomness it sent and either outputs an Accept or Reject.

# Updated

We would like now to move to the following picture

![proposal_aux_trace](https://github.com/facebook/winterfell/assets/82364884/f22ad592-edc4-4bf5-a208-5fea4136b8cc)

The main difference now is that we allow the auxiliary trace building stage to be potentially more complex than just the verifier sending randomness and the prover answering with commitments to the auxiliary trace columns.
More specifically, in the case of GKR-LogUp, we could have the following simplified flow

![gkr_builder_aux_trace](https://github.com/facebook/winterfell/assets/82364884/50a7fd8b-a618-4723-be25-d63ec961a320)

1. The verifier sends $\alpha := \left(\alpha_i\right)_i$ i.e., LogUp randomness.
2. Both the prover and verifier compute $C\left\(\alpha\right)$.
3. GKR is used for the statement $$\sum_{x\in\\{0,1\\}^{\nu}} \frac{p(x)}{q(x)} = C\left\(\alpha\right).$$ Note that we abuse the notation with regards to MLEs . Note also that $p$ and $q$ depend on $\alpha$ implicitly.
4. After a certain number of rounds making up the GKR IOP, the verifier is left with a GKR proof $\mathsf{\pi_{gkr}}$ and the prover is left with a vector of random values $\mathsf{r} := \left(\mathsf{r_i}\right)_i$.
5. The prover sends $\mathsf{y_i}$ it claims are the evaluations of the MLEs of $\mathsf{f_i}$.
6. The verifier sends randomness $\lambda := \left(\lambda_i\right)_i$ in order to batch the (claimed) evaluations into one evaluation $\mathsf{y} := \sum_i \mathsf{y_i}\cdot\lambda_i$.
7. The prover sends $\mathsf{commit(l)}$ and $\mathsf{commit(s)}$ where:  
           1. $\mathsf{l}$ is the Lagrange kernel at $\mathsf{r}$.
           2. $\mathsf{s}$ is the univariate sum-check column i.e., a column such that $\mathsf{s[row]} = \mathsf{s[row-1]} + \sum_i \lambda_i\mathsf{f_i(row)}\cdot\mathsf{l(row)}$[^1].
8. Both the prover and verifier update[^2] their AIR instance using $\mathsf{r}$, $\lambda$ and $\mathsf{y}$. This includes a boundary constraint against the last row of column $\mathsf{s}$ forcing it to be equal to $\mathsf{y}$ as well as the transition constraint in point 7.2 above. There are also the Lagrange kernel constraints that are defined using $\mathsf{r}$.

To summarize, with the new changes, the auxiliary trace building stage would culminate in a proof, an auxiliary segment composed of two columns and an updated AIR instance. 
As can be seen, there is a strong similarity between the above and the FRI stage i.e., both are IOPs with a non-constant number of rounds.
The task now is to discuss possible ways we can accommodate the above with the least amount of abstraction necessary i.e., offloading most of the work which happens at the auxiliary trace building stage to the "app" instead of the proof system.

[^1]: The sum might also include shifts of $\mathsf{f_i}$ as well as periodic columns. These are omitted in the analysis above for simplicity.
[^2]: We can also think of some of the constraint as being uninitialized and they get initialized using the appropriate randomness.

**Comment by plafer** [comment]
I believe the following would work, and doesn't require "updating the AIR instance" (as mentioned in the diagram).

We add a `Prover:AuxTraceBuilder: AuxTraceBuilder` associated type, where

```rs
trait AuxTraceBuilder {
    // Default: Vec<E>
    // GKR: GkrRandElements<E>
    type AuxRandEles<E>;
    // Default: num_rand_elements
    // GKR: ()
    type AuxParams;
    // Default: ()
    // GKR: GkrProof
    type AuxProof;

    fn build_aux_trace<E: FieldElement>(
        &mut self,
        main_trace: &ColMatrix<E::BaseField>,
        aux_params: Self::AuxParams,
        transcript: &mut PublicCoin,
    // Returns
    // - aux trace
    // - random elements: Necessary for constraint evaluation (e.g. `Air::evaluate_aux_transition()`)
    // - aux proof: potentially a GKR proof
    ) -> (ColMatrix<E>, Self::AuxRandEles<E>, Option<Self::AuxProof>);
}

/// The `AuxTraceBuilder::RandEles` that the GKR-based implementation would use
struct GkrRandElements<E> {
    input_layer_rands: Vec<E>,
    lambda: Vec<E>,
}
```

Hence, the `Trace` is no longer responsible for building the auxiliary trace (which doesn't make sense for the GKR-based implementation). 

Similarly, we'd remove `TraceInfo::num_aux_segment_rands`, which doesn't make sense for GKR (since the number of auxiliary random elements isn't fixed). This is replaced by `AuxTraceBuilder::AuxParams`: we'd pass `num_aux_segment_rands` explicitly for the default implementation (i.e. the current one). Hence, `Prover::prove<AuxParams>(trace, aux_params: AuxParams)` now needs that extra `aux_params` argument.

Finally, we'd also add an `Air: AuxRandEles` associated type, which would be used in the auxiliary constraint evaluation methods:

+ `get_lagrange_rand_eles(rand_eles: &Self::AuxRandEles) -> &[E]`
    + selects the Lagrange rand eles from its `Self::AuxRandEles`
+ `evaluate_aux_transition(..., Self::AuxRandEles)`
    + Now passes the associated `AuxRandEles` that contains the lambdas (in GKR case)
+ `get_aux_assertions(Self::AuxRandEles)`
    + Can access the lambdas as well (in GKR case)

**Comment by irakliyk** [comment]
Closed by #271 - thought we probably will need to refactor this a bit when implementing GKR support more concretely.
